oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrantBuilder.java

```diff
@@ -0,0 +1,198 @@
+        this.accessTokenEndpointPath = requireNonNull(accessTokenEndpointPath, "accessTokenEndpointPath");
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on encoded authorization token and
+     * authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param authorizationSupplier A supplier of encoded client authorization token.
+     * @param authorizationType One of the registered HTTP authentication schemes as per
+     *                          <a href="https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml">
+     *                          HTTP Authentication Scheme Registry</a>.
+     */
+    @SuppressWarnings("unchecked")
+    public final T clientAuthorization(
+            Supplier<String> authorizationSupplier, String authorizationType) {
+        clientAuthorization = ClientAuthorization.ofAuthorization(authorizationSupplier, authorizationType);
+        return (T) this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on encoded authorization token and
+     * {@code Basic} authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param authorizationSupplier A supplier of encoded client authorization token.
+     */
+    @SuppressWarnings("unchecked")
+    public final T clientBasicAuthorization(Supplier<String> authorizationSupplier) {
+        clientAuthorization = ClientAuthorization.ofBasicAuthorization(authorizationSupplier);
+        return (T) this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on client credentials and
+     * authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param credentialsSupplier A supplier of client credentials.
+     * @param authorizationType One of the registered HTTP authentication schemes as per
+     *                          <a href="https://www.iana.org/assignments/http-authschemes/http-authschemes.xhtml">
+     *                          HTTP Authentication Scheme Registry</a>.
+     */
+    @SuppressWarnings("unchecked")
+    public final T clientCredentials(
+            Supplier<? extends Map.Entry<String, String>> credentialsSupplier, String authorizationType) {
+        clientAuthorization = ClientAuthorization.ofCredentials(credentialsSupplier, authorizationType);
+        return (T) this;
+    }
+
+    /**
+     * Provides client authorization for the OAuth 2.0 requests based on client credentials and
+     * {@code Basic} authorization type,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     *
+     * @param credentialsSupplier A supplier of client credentials.
+     */
+    @SuppressWarnings("unchecked")
+    public final T clientCredentials(
```

Comment:
no, because there are userCredentials as well. here "client" is the OAuth2 spec term, which represents client application/device/etc vs. "user" which represents a human being

Reply:
Got it. Thanks.